src/services/highlight.tsx

```diff
@@ -0,0 +1,228 @@
+  let lineFilterMatchIndex = 0;
+  let matchInterval = matchingIntervals[lineFilterMatchIndex];
+
+  // @todo debug why it's only grabbing the first one
+  for (let valueIndex = 0; valueIndex < value.length; valueIndex++) {
+    // Size is 1 based length, lineFilterMatchIndex is 0 based index
+    while (valueIndex >= matchInterval[1] && lineFilterMatchIndex < size - 1) {
+      lineFilterMatchIndex++;
+      matchInterval = matchingIntervals[lineFilterMatchIndex];
+    }
+    if (valueIndex >= matchInterval[0] && valueIndex < matchInterval[1]) {
+      // this char is part of highlight, return an object in the array so we don't lose the original order, and we can differentiate between highlighted text in the subsequent merge
+      valueArray.push({ value: value[valueIndex] });
+    } else {
+      valueArray.push(value[valueIndex]);
+    }
+  }
+
+  return mergeStringsAndElements(valueArray);
+};
+
+// @todo cache results by regex/value?
+export const getLineFilterMatches = (matchExpressions: RegExp[], value: string): Array<[number, number]> => {
+  let results: Array<[number, number]> = [];
+  matchExpressions.forEach((regex) => {
+    let valueMatch;
+    let valueMatches = [];
+    do {
+      try {
+        valueMatch = regex.exec(value);
+        if (valueMatch) {
+          valueMatches.push(valueMatch);
+        }
+      } catch (e) {
+        logger.error(e, { msg: 'Error executing match expression', regex: regex.source });
+      }
+    } while (valueMatch);
+    if (valueMatches.length) {
+      const fromToArray: Array<[number, number]> = valueMatches.map((vm) => [vm.index, vm.index + vm[0].length]);
+      results.push(...fromToArray);
+    }
+  });
+
+  return results;
+};
+
+function mergeOverlap(arr: number[][]) {
+  // Merge overlapping intervals in-place. We return
+  // modified size of the array arr.
+
+  // Sort intervals based on start values
+  arr.sort((a, b) => a[0] - b[0]);
+
+  // Index of the last merged
+  let resIdx = 0;
+
+  for (let i = 1; i < arr.length; i++) {
+    // If current interval overlaps with the
+    // last merged interval
+    if (arr[resIdx][1] >= arr[i][0]) {
+      arr[resIdx][1] = Math.max(arr[resIdx][1], arr[i][1]);
+    }
+    // Move to the next interval
+    else {
+      resIdx++;
+      arr[resIdx] = arr[i];
+    }
+  }
+
+  // Returns size of the merged intervals
+  return resIdx + 1;
+}
+
+export const mergeOverlapping = (matchIndices: number[][]) => {
+  if (matchIndices.length) {
+    return mergeOverlap(matchIndices);
+  }
+  return 0;
+};
+
+export const getLogsHighlightStyles = (theme: GrafanaTheme2, showHighlight: boolean) => {
+  if (!showHighlight) {
+    return {};
+  }
+
+  // @todo find way to sync/pull from core?
+  const colors = {
+    critical: '#B877D9',
+    debug: '#6E9FFF',
+    error: theme.colors.error.text,
+    info: '#6CCF8E',
+    metadata: theme.colors.text.primary,
+    parsedField: theme.colors.text.primary,
+    trace: '#6ed0e0',
+    warning: '#FBAD37',
```

Comment:
I believe warning is `theme.colors.warning.text` but yeah not sure about critical etc